### ISSUE TITLE
Provide JSON formatting option for layout=all-event-properties

### ISSUE THREAD

**Issue by jefflowe7** [NLog/NLog#1686]
**Type**: Feature request
**NLog version**: 4.3.5
**Platform**: .Net 4.5 

My request seems to already exist at https://github.com/NLog/NLog/issues/723, but it doesn't appear to be resolved.

Many logging services (Sumologic, Kibana) provide a robust JSON parsing capability. When logging an event, much important information is contained in the Properties on the event. However it doesn't appear that those Properties can be formatted as JSON via NLog.

Here's the layout config:

``` xml
  <layout xsi:type="JsonLayout">
      <attribute name="time" layout="${shortdate}" />
      <attribute name="level" layout="${level:upperCase=true}" />
      <attribute name="message" layout="${message}" />
      <attribute name="properties" layout="${all-event-properties}" />
  </layout>
```

Here's an example NLog call:

``` c#
    Exception e = new Exception("Something went wrong.");

    string serializedException = JsonSerialize(e);

    Logger
        .Info()
        .Message("Start.")
        .Property("Exception", serializedException)
        .Property("Host", "http://xyzzy.gov")
        .Write();
```

Here's an example of the output using the layout above:

``` javascript
  {
    "time": "2015-05-30 17:11:50.7131",
    "level": "INFO",
    "message": "Start",
    "properties": "Exception={"Message": "Something went wrong."}, Host="http://xyzzy.gov""
  }
```

Note that the value of properties is NOT JSON.

What we need is:

``` javascript
  {
    "time": "2015-05-30 17:11:50.7131",
    "level": "INFO",
    "message": "Start",
    "properties": {
        "Exception": {
            "Message": "Something went wrong"
        }
        "Host": "http://xyzzy.gov"
      }
  }
```

Note that for the Exception property, the property value as already been logged as serialized JSON.

**Comment by 304NotModified** [comment]
> Note that the value of properties is NOT JSON.

Agree, this isn't usefull. What is the result with `encode="false"`, e.g.

```xml
  <layout xsi:type="JsonLayout">
      <attribute name="time" layout="${shortdate}" />
      <attribute name="level" layout="${level:upperCase=true}" />
      <attribute name="message" layout="${message}" />
      <attribute name="properties" layout="${all-event-properties}" encode="false" />
  </layout>
```

**Comment by jefflowe7** [comment]
Tried:

'''xml
    <attribute name="properties" layout="${all-event-properties}" encode="false" />
'''
Still getting:

'''
{ 
    "time": "2016-10-06", 
    "level": "DEBUG", 
    "message": "Sending request.", 
    "properties": Request={"AlwaysMultipartFormData":false,"JsonSerializer": ...
}

'''

Still the same problem.

**Comment by 304NotModified** [comment]
Indeed.  any proposal how to fix this? (New setting? Where?)

**Comment by jefflowe7** [comment]
Same result, e.g.:
{   "time": "2016-10-06",   "level": "DEBUG",   "message": "Sending request.",   "properties": Request={"AlwaysMultipartFormData":false,"JsonSerializer": etc.}

**Comment by 304NotModified** [comment]
Hi, this is the same response? 

I'm curious how to configure this, as it is preferred that the all-event-properties-layoutrenderer isn't aware that it's writing Json

**Comment by aireq** [comment]
I would treat this like how the [Elasticsearch NLog target](https://github.com/ReactiveMarkets/NLog.Targets.ElasticSearch/wiki) handles this with `includeAllProperties` and `excludedProperties`.


An empty `JsonLayout`  would still generate nothing or empty objects as it already does.

    <layout xsi:type="JsonLayout" />

Adding `includeAllProperties` would automatically write all the properties of each log event to corresponding properties of the JSON object.

    <layout xsi:type="JsonLayout" includeAllProperties="true"/>


Adding `excludedProperties` would allow specific properties to be omitted
    <layout xsi:type="JsonLayout" includeAllProperties="true" excludedProperties="BadPropertyName1,BadProperty2"/>


`field` elements could still be added for properties that are added automatically. This would just overwrite their values with the layout of the field.

    <layout xsi:type="JsonLayout" includeAllProperties="true" excludedProperties="BadPropertyName1, BadProperty2">
        <field name="message" layout="My message for you is this ${message}" />
    </layout>

**Comment by 304NotModified** [comment]
@aireq ! good proposal! 

This isn't that difficult to implement. :)

If we point you to the right direction, could you code this and send a PR?

**Comment by aireq** [comment]
Sure I can take a stab at it.

**Comment by 304NotModified** [comment]
Cool!

**Comment by 304NotModified** [comment]
Fixed by https://github.com/NLog/NLog/pull/1754

will be released in 4.4

**Comment by 304NotModified** [comment]
Fixed so close :)
